cmd/sprint-automation/main.go

```diff
@@ -386,6 +447,112 @@ func sendIntakeDigest(slackClient *slack.Client, jiraClient *jiraapi.Client, use
+		stable:  current.State == configv1.CompletedUpdate && current.CompletionTime != nil && time.Since(current.CompletionTime.Time) > 24*time.Hour,
+	}
+	cv, err := semver.Make(current.Version)
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine sematic version: %s", current.Version)
+	}
+	ret.semanticVersion = cv
+	if ret.stable && len(status.History) > 1 {
+		previous := status.History[1]
+		pv, err := semver.Make(previous.Version)
+		if err != nil {
+			return nil, fmt.Errorf("failed to determine sematic version: %s", previous.Version)
+		}
+		if cv.Minor > pv.Minor {
+			// Need to soak a week after a Y-stream upgrade
+			ret.stable = time.Since(current.CompletionTime.Time) > 7*24*time.Hour
+		}
+	}
+	return ret, nil
+}
+
+func clusterVersion(ctx context.Context, clusterName string, Client ctrlruntimeclient.Reader) (*versionInfo, error) {
+	cv := &configv1.ClusterVersion{}
+	if err := Client.Get(ctx, ctrlruntimeclient.ObjectKey{Name: "version"}, cv); err != nil {
+		return nil, fmt.Errorf("failed to get ClusterVersion version on %s: %w", clusterName, err)
+	}
+	return newVersionInfo(cv.Status)
+}
+
+func upgradeBuild02(ctx context.Context, build01Client, build02Client ctrlruntimeclient.Reader) (string, error) {
+	build01VI, err := clusterVersion(ctx, "build01", build01Client)
+	if err != nil {
+		return "", err
+	}
+	if !build01VI.stable {
+		logrus.WithField("build01Version", build01VI.version).Info("The version on build01 has not been stable enough and hence no need to upgrade build02")
+		return "", nil
+	}
+
+	build02VI, err := clusterVersion(ctx, "build02", build02Client)
+	if err != nil {
+		return "", err
+	}
+	if build02VI.state != configv1.CompletedUpdate {
+		logrus.WithField("state", build02VI.state).Info("The previous upgrade of build02 has not been completed")
+		return "", nil
+	}
+	if build02VI.semanticVersion.Equals(build01VI.semanticVersion) {
+		logrus.WithField("version", build01VI.version).Info("build01 and build02 have the same version and hence no need to upgrade build02")
+		return "", nil
+	}
+
+	if build02VI.semanticVersion.GT(build01VI.semanticVersion) {
+		return "", fmt.Errorf("version of build02 %s is newer than build01 %s", build02VI.version, build01VI.version)
+	}
+	return build01VI.version, nil
+}
+
+func sendTriageBuild02Upgrade(slackClient *slack.Client, version string) error {
+	blocks := []slack.Block{
+		&slack.HeaderBlock{
+			Type: slack.MBTHeader,
+			Text: &slack.TextBlockObject{
+				Type: slack.PlainTextType,
+				Text: "Upgrade Build02",
+			},
+		},
+		&slack.SectionBlock{
+			Type: slack.MBTSection,
+			Text: &slack.TextBlockObject{
+				Type: slack.PlainTextType,
+				// Ideally, we could just run the upgrade command and notify triage via slack
+				// In reality we still need some manual checks before upgrading build02
+				Text: fmt.Sprintf("@%s Please upgrade `build02` if `build02` is healthy: `oc --as system:admin --context build02 adm upgrade --to=%s`", userGroupTriage, version),
```

Comment:
I think we could include more context: "Version X is stable on b01 for a week. Please upgrade b02 to that version by running --...". Maybe add a link to triage doc.

Reply:
Updated.